### PR TITLE
Endpoint to list entities needing linking.

### PR DESCRIPTION
**Pull request by jokedurnez**

### CODE REVIEW

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 1]
```diff
@@ -52,6 +51,23 @@ class RealtimeLinkingController(
         return linkableEntitySets.minus(entitySetsNeedLinking)
     }
 
+    @RequestMapping(
+            path = [RealtimeLinkingApi.MISSING],
+            method = [RequestMethod.POST],
+            produces = [MediaType.APPLICATION_JSON_VALUE]
+    )
+    override fun getEntitiesMissingLinking(@RequestBody entitySetIds: Set<UUID>): Map<UUID, Set<UUID>> {
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: You don't use the parameter

Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: I would check, if the requested entity set ids are actually linkable entity sets and either throw `IllegalArgumentException` if we want to avoid calls like that or log a warning with the non-linkable entity set ids.
For that you will need to add hazelcastinstance to the constructor and get the entity sets map. I already did that for different reasons in this PR: https://github.com/openlattice/linker/pull/52/files You can copy those parts over.
To get the non-linkable entity set ids you could do something like this:
```
val linkableEntitySetIds = lqs
                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
                .toSet()
val nonLinkableEntitySetIds = entitySetIds.filter { !linkableEntitySets.contains(it) }
```
If you decide to throw error then:
```
require(nonLinkableEntitySetIds.isEmpty()) {"Some error message with $nonLinkableEntitySetIds"}
```

If not you can just log it:
```
if(nonLinkableEntitySetIds.isNotEmpty()) {
    logger.warn("Message with $nonLinkableEntitySetIds")
}
```

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 2]
```diff
@@ -52,6 +51,23 @@ class RealtimeLinkingController(
...
+        ensureAdminAccess()
+        val linkableEntitySets = lqs
+                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
+                .toSet()
+        val entitiesNeedLinking = HashMap<UUID, Set<UUID>>()
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: Why not do a batch call? You load those UUIDs anyway into memory

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 1 above)
Comment (you-leee) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: An alternative could be to make a function in `LinkingQueryService` that returns the linkable entity set ids from requested entity set ids (and not all)

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 3]
```diff
@@ -52,6 +51,23 @@ class RealtimeLinkingController(
...
+                linkableEntitySets.forEach {
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: This should query the requested entity set ids and not all linkable entity set ids

File: src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package com.openlattice.linking;
+
+public class LinkingScoringTest {
```
Comment (you-leee) [inline src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java:3]: I don't think you can use rehearsal functions here if you want to load the data first and then do the load test here Assuming this test will be for that, if it tests only the scores/is a unit test then just ignore this comment :)

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 1 above)
Comment (jokedurnez) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: Why not:
```
    override fun getEntitiesMissingLinking(@RequestBody entitySetIds: Set<UUID>): Map<UUID, Set<UUID>> {
        ensureAdminAccess()
        val linkableEntitySets = lqs
                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
                .toSet()
                .intersect(entitySetIds)
```
Taking the intersection between `linkableEntitySets` and `entitySetIds`?

File: src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java (hunk 4 above)
Comment (jokedurnez) [reply src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java:3]: Yeah, I maybe should remove this?
It should be a unit test in the future to see if the input array of the features results in the correct score, to make sure the training order/features/... map to prod.

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 1 above)
Comment (you-leee) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: That's fine for getting the entitySetIds to pass to `lqs.getEntitiesNotLinked`, but I would still at least log entity set ids, that are non-linkable from the input ids

File: src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java (hunk 4 above)
Comment (you-leee) [reply src/test/kotlin/com/openlattice/linking/LinkingScoringTest.java:3]: You can keep it here just add a `// TODO It should be a unit test in the future to see if the input array of the features results in the correct score, to make sure the training order/features/... map to prod`

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 2 above)
Comment (you-leee) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: `lqs.getEntitiesNotLinked(linkableEntitySets)
    .groupBy { it.first }
    .mapValues { it.value.map { it.second }.toSet() }`

Comment (you-leee) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: ^ this will map it to the desired output type: First you group the pairs by entity set id (`it.first`) , which results in `Map<UUID, List<Pair<UUID, UUID>>>` and then you map this maps values

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 1 above)
Comment (jokedurnez) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: 🤦‍♀

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 5]
```diff
@@ -52,6 +56,33 @@ class RealtimeLinkingController(
...
+    override fun getEntitiesMissingLinking(@RequestBody entitySetIds: Set<UUID>): Map<UUID, Set<UUID>> {
+        ensureAdminAccess()
+        val linkableEntitySets = lqs
+                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
+                .toSet()
+        val linkableRequestedEntitySets = linkableEntitySets
+                .intersect(entitySetIds)
+        val nonLinkableRequestedEntitySets = entitySetIds.subtract(linkableEntitySets)
+        if(nonLinkableRequestedEntitySets.isNotEmpty()) {
+            logger.warn("The following entitySets were queried for missing entities, " +
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: I would rephrase that to "for entities missing linking id" or "for entities not yet linked" missing entities is a bit misleading

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 6]
```diff
@@ -52,6 +56,33 @@ class RealtimeLinkingController(
...
+        val linkableRequestedEntitySets = linkableEntitySets
+                .intersect(entitySetIds)
+        val nonLinkableRequestedEntitySets = entitySetIds.subtract(linkableEntitySets)
+        if(nonLinkableRequestedEntitySets.isNotEmpty()) {
+            logger.warn("The following entitySets were queried for entities not yet linked, " +
+                    "but are not linkable: $nonLinkableRequestedEntitySets")
+        }
+
+        val entitiesNeedLinking = lqs
+                .getEntitiesNotLinked(linkableRequestedEntitySets, 100000000)
```
Comment (pckeyes) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: We lean towards `100_000_000` rather than `100000000` for readability

Comment (jokedurnez) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: oOoOo interesting !

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 7]
```diff
@@ -52,6 +56,33 @@ class RealtimeLinkingController(
...
+                .toSet()
+        val linkableRequestedEntitySets = linkableEntitySets
+                .intersect(entitySetIds)
+        val nonLinkableRequestedEntitySets = entitySetIds.subtract(linkableEntitySets)
+        if(nonLinkableRequestedEntitySets.isNotEmpty()) {
+            logger.warn("The following entitySets were queried for entities not yet linked, " +
+                    "but are not linkable: $nonLinkableRequestedEntitySets")
+        }
+
+        val entitiesNeedLinking = lqs
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: Really minor, but you could just return the result of this and not assign it to a variable

**Review by you-leee — APPROVED** [review]

**Review by makosblade — CHANGES_REQUESTED** [review]
see: https://github.com/openlattice/api/pull/416#pullrequestreview-365101270

**Comment by jokedurnez** [comment]
Let's keep the discussion in api.
Changed into count.

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 8]
```diff
@@ -52,6 +56,30 @@ class RealtimeLinkingController(
...
+    override fun getEntitySetNotYetLinkedCount(@RequestBody entitySetIds: Set<UUID>): Map<UUID, Long> {
+        ensureAdminAccess()
+        val linkableEntitySets = lqs
+                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
+                .toSet()
+        val linkableRequestedEntitySets = linkableEntitySets
+                .intersect(entitySetIds)
+        val nonLinkableRequestedEntitySets = entitySetIds.subtract(linkableEntitySets)
+        if(nonLinkableRequestedEntitySets.isNotEmpty()) {
+            logger.warn("The following entitySets were queried for entities not yet linked, " +
```
Comment (dbt) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: Users sending bad inputs that result in logs is usually not a good idea.

This should either tell the user or do nothing.

(Under some circumstances using a counter would be valuable, I don't think that's true here).

**Review by dbt — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt (hunk 8 above)
Comment (geekbeast) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: Good point this warning doesn't seem to be make a lot of sense.

**Review by geekbeast — CHANGES_REQUESTED** [review]

**Comment by jokedurnez** [comment]
removed the warning.

File: src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt [hunk 9]
```diff
@@ -52,6 +56,26 @@ class RealtimeLinkingController(
...
+    )
+
+    override fun getEntitySetNotYetLinkedCount(@RequestBody entitySetIds: Set<UUID>): Map<UUID, Long> {
+        ensureAdminAccess()
+        val linkableEntitySets = lqs
+                .getLinkableEntitySets(linkableTypes, entitySetBlacklist, whitelist.orElse(setOf()))
+                .toSet()
+        val linkableRequestedEntitySets = linkableEntitySets
+                .intersect(entitySetIds)
+        val nonLinkableRequestedEntitySets = entitySetIds.subtract(linkableEntitySets)
```
Comment (katherinebernstein) [inline src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: What is `nonLinkableRequestedEntitySets` used for?

Comment (jokedurnez) [reply src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt]: it's removed
